src/analyzing/AST_to_IL.ml

```diff
@@ -685,6 +685,52 @@ and expr_aux env ?(void = false) e_gen =
         | _ -> exp
       in
       ident_function_call_hack exp
+  (* x = ClassName(args ...) in Python *)
+  (* ClassName has been resolved to __init__ by the pro engine. *)
+  (* Identified and treated as x = New ClassName(args ...) to support
+     field sensitivity. *)
+  | G.Assign
+      ( ({
+           e =
+             ( G.N (G.Id ((_, _), lhs_info))
+             | G.DotAccess (_, _, FN (Id ((_, _), lhs_info))) );
+           _;
+         } as lhs_e),
+        _,
+        ({
+           e =
+             G.Call
+               ( {
+                   e =
+                     ( G.N (Id (_, id_info))
+                     | G.DotAccess (_, _, FN (Id (_, id_info))) );
+                   _;
+                 },
+                 args );
+           _;
+         } as origin_exp) )
+  (* Can we be tighter about side conditions here to avoid catching
+     direct calls to __init__ ? *)
+    when match id_info.G.id_resolved.contents with
+         | Some (G.GlobalName (ls, _), _) -> (
+             env.lang =*= Lang.Python
+             && List.length ls >= 3 (* Module + Class + __init__ *)
+             && (match List_.last_opt ls with
+                | Some "__init__" -> true
+                | _ -> false)
+             &&
+             match lhs_info.id_type.contents with
+             | Some { t = G.TyN _; _ } -> true
+             | _ -> false)
+         | _ -> false ->
+      let lval = lval env lhs_e in
+      let ty =
+        match lhs_info.id_type.contents with
+        | Some ty -> ty
+        | _ -> failwith "Impossible: guard above checks that this exists"
```

Comment:
I'd rather pattern match on `lhs_info` and have the pattern match duplicated in both cases where that is bound than have a failwith here.

Reply:
Definitely no `failwith` allowed...